### PR TITLE
added cylc get-resources --tutorials

### PR DESCRIPTION
**Pull request by wxtim**
These changes close https://github.com/cylc/cylc-doc/issues/40

The intention with this is that the documentation can set out the following workflow to users:

```
cylc get-resources --tutorials
cd ~/cylc-src/cylc-tutorials/<tutorial-name>

# ... make whatever changes the tutorial requires...

cylc install cylc-tutorials/<tutorial-name>
cylc validate cylc-tutorials/<tutorial-name>
cylc play cylc-tutorials/<tutorial-name>
```

**Requirements check-list**
- [x] I have read `CONTRIBUTING.md` and added my name as a Code Contributor.
- [x] Contains logically grouped changes (else tidy your branch by rebase).
- [x] Does not contain off-topic changes (use other PRs for other changes).
- [x] Applied any dependency changes to both `setup.cfg` and `conda-environment.yml`.
- [x] Appropriate tests are included (unit and/or functional).
- [ ] Appropriate change log entry included.
- [x]  Comprehensive documentation update required: This ticket should be merged first.

### CODE REVIEW

File: cylc/flow/resources.py [hunk 1]
```diff
@@ -61,7 +70,10 @@ def get_resources(target_dir, resources=None):
     if resources is None:
         resources = resource_names
     for resource in resources:
-        if resource not in resource_names:
+        if (
+            resource not in resource_names
+            and resource != 'etc/tutorial'
+        ):
```
Comment (MetRonnie) [inline cylc/flow/resources.py:76]: Why not just add `'etc/tutorial'` to `resource_names`?

File: cylc/flow/resources.py [hunk 2]
```diff
@@ -72,3 +84,20 @@ def get_resources(target_dir, resources=None):
         res = pr.resource_string('cylc.flow', resource)
         with open(path, 'wb') as h:
             h.write(res)
+
+
+def extract_tutorials():
+    """Extract tutorial workflows to Cylc Source Directory."""
+    src = Path(cylc.flow.__file__).parent / 'etc/tutorial'
+    dest = glbl_cfg().get(['install', 'source dirs'])[0]
```
Comment (MetRonnie) [inline cylc/flow/resources.py:92]: What if no source dirs are configured?

File: tests/unit/test_resources.py [hunk 3]
```diff
@@ -60,3 +62,20 @@ def test_cli(tmpdir):
...
         )
+
+
+def test_extract_tutorials(monkeypatch, tmp_path, caplog):
+    test_dest = tmp_path/ 'destination'
+    class MockGlobalConfig:
+        def get(*args):
+            return [str(test_dest)]
+
+    monkeypatch.setattr(GlobalConfig, '_DEFAULT', MockGlobalConfig)
```
Comment (MetRonnie) [inline tests/unit/test_resources.py:73]: You could also use the `mock_glbl_cfg` fixture
```
mock_glbl_cfg(
    'cylc.flow.resources.glbl_cfg',
    f'''
    [install]
        source dirs = {test_dest}
     '''
)
```

File: cylc/flow/resources.py (hunk 2 above)
Comment (wxtim) [reply cylc/flow/resources.py:92]: Doesn't it have a default value in the config of `$HOME/cylc-src`?

File: tests/unit/test_resources.py (hunk 3 above)
Comment (wxtim) [reply tests/unit/test_resources.py:73]: Thanks - I thought that such a thing existed, but I was clearly not thinking straight at the time I wrote this.

File: cylc/flow/resources.py (hunk 1 above)
Comment (wxtim) [reply cylc/flow/resources.py:76]: Because this function handles files, but not directories.

Comment (oliver-sanders) [reply cylc/flow/resources.py:76]: We could make the main function support directories and make `~/cylc-src` the default dest directory for resources in the tutorial function. That way users wouldn't have to type the `--tutorial` option and the `[DIR]` positional argument would still work.

**Review by MetRonnie — APPROVED** [review]

File: cylc/flow/resources.py (hunk 1 above)
Comment (wxtim) [reply cylc/flow/resources.py:76]: OK, I'll do that then.

**Review by hjoliver — CHANGES_REQUESTED** [review]
I think it's fine to treat the tutorials differently from other resources, but we should update the documentation accordingly.

- This doesn't list the tutorials:
```
  --list                List available package resources.`
```
- This doesn't extract the tutorials:
```Arguments:
   [RESOURCES...]               Resources to extract (default all).
```
- This should say *to ~/cylc-src/cylc-tutorials*?
```
  --tutorials           Extract tutorials workflows to ~/cylc-src
```
- And maybe `cylc get-resources --tutorials` should also print "extracted to ~cylc-src/cylc-tutorials`?

**Comment by wxtim** [comment]
@oliver-sanders I think that I have responded to Hilary's review comments. If you are particularly keen to merge you could check that.

**Review by oliver-sanders — CHANGES_REQUESTED** [review]
Unfortunately I may have forgotten to mention an undocumented feature of `rose tutorial`, that replaced API keys in the tutorials automatically. My bad, I should have spotted this sooner.

**Review by oliver-sanders — APPROVED** [review]

File: cylc/flow/resources.py [hunk 4]
```diff
@@ -14,99 +14,180 @@
...
+                            b'DATAPOINT_API_KEY',
+                            api_key.encode(),
+                        )
+                    )
+
+            # then move the tmpfile over the original
+            # NOTE: shutil interfaces don't fully support Path objects at all
+            # python versions
+            path.unlink()
+            shutil.move(str(tmp_path), str(path))
```
Comment (MetRonnie) [inline cylc/flow/resources.py:193]: Why not simplify this
```suggestion
            with open(path, 'rb') as f:
                # NOTE: open the file in bytes mode for safety
                # (prevents decode errors surfacing here)
                contents = f.read()
            contents = contents.replace(
                b'DATAPOINT_API_KEY',
                api_key.encode()
            )
            with open(path, 'wb') as f:
                f.write(contents)
```

File: cylc/flow/scripts/get_resources.py [hunk 5]
```diff
@@ -17,14 +17,34 @@
 """cylc get-resources [OPTIONS] ARGS
 
 Extract resources from the cylc.flow package.
-"""
+
+If the directory is omitted the resource will be coppied either to your
```
Comment (MetRonnie) [inline cylc/flow/scripts/get_resources.py:21]:
```suggestion
If the directory is omitted the resource will be copied either to your
```

File: cylc/flow/scripts/get_resources.py [hunk 6]
```diff
@@ -17,14 +17,34 @@
...
+current working directory for regular resources, or to the first configured
+"source" directory (~/cylc-src by default).
```
Comment (MetRonnie) [inline cylc/flow/scripts/get_resources.py:23]: I think?
```suggestion
"source" directory (~/cylc-src by default) for tutorials.
```

File: cylc/flow/resources.py (hunk 4 above)
Comment (oliver-sanders) [reply cylc/flow/resources.py:193]: More efficient reading/writing one line at a time.

**Pull request merged**